### PR TITLE
Feature/cursor move by grapheme

### PR DESCRIPTION
**Pull request by gpwclark**
This just builds off my previous PR. But uh, is ridiculously faster.

### CODE REVIEW

File: src/buffer.rs [hunk 1]
```diff
@@ -16,32 +19,34 @@ impl Action {
...
             Action::Insert { start, ref text } => {
-                buf.insert_raw(start, &text[..]);
+                buf.insert_raw(start, text);
                 Some(start)
             }
             Action::Remove { start, ref text } => {
-                let text_len = text.len();
-                buf.remove_raw(start, start + text_len).len();
-                if text_len > start {
+                let len = text.len();
```
Comment (gpwclark) [inline src/buffer.rs:26]: Width? Not 3ln?

**Review by sstanfield — COMMENTED** [review]
General comment, I don't feel strongly about this, it may be impractical (&str vs String issues) or too much trouble for too little gain.  Maybe consider making a new type for chars/grapheme clusters just to make it obvious what is being passed.  Again, not sure that is remotely worth the effort it would take and knowing a 'char' is a 'string' is just something someone should know to use liner now.  Once lifetimes get added to the mix the syntax/permutations would likely be unworkable (I have not thought that threw).

File: src/grapheme_iter.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+                // somewhere in the middle
+                let end = self.offsets[self.curr_grapheme];
+                str = Some(&self.data[start..end]);
+            }
+        }
+        str
+    }
+}
+
+impl<'a> DoubleEndedIterator for GraphemeIter<'a> {
```
[the remark this answers is not in the document]
Comment (gpwclark) [reply src/grapheme_iter.rs:157]: was able to simplify these and fix this issue.

File: src/grapheme_iter.rs [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+        }
+        str
+    }
+}
+
+impl<'a> Read for GraphemeIter<'a> {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let len = buf.len() - 1;
+        let mut bytes = self.data.as_bytes()[..len].to_owned();
+        buf.swap_with_slice(&mut bytes);
```
[the remark this answers is not in the document]
Comment (gpwclark) [reply src/grapheme_iter.rs]: I wrote some test cases, grapheme_iter.rs 379-419

I checked the docs for Read. Do these test cases match up with your understanding?

[the remark this answers is not in the document]
Comment (sstanfield) [reply src/grapheme_iter.rs]: You code (and test cases) seem to assume the read returns Ok(n) where n = the size of buf but if there is less data available then the it should be Ok(n) where n = the bytes actually copied.  This discussion is out of sync with code but the fix looks trivial by moving the return into the if and setting the other case correctly.

[the remark this answers is not in the document]
Comment (sstanfield) [reply src/grapheme_iter.rs]: Also test a iter with no data and make sure that is ok.  You may need to special case that.

[the remark this answers is not in the document]
Comment (gpwclark) [reply src/grapheme_iter.rs]: cool, I think i handle all 3 cases now. tarpaulin was knocking me for not having some tests on some empty grapheme iterators so I knocked out that special case w/ read and the rest of the public methods on an empty grapheme iter as well.

[the remark this answers is not in the document]
Comment (sstanfield) [reply src/grapheme_iter.rs]: So, a little late on this comment and also possibly off base (I have not looked at why you implemented Read just assumed you had a good reason) but would something like String::as_bytes() serve the purpose better?  Again, have not looked at the code that uses Read but if you added as_bytes (basically a pass through to the underlying string) that may do the same thing without needing a slice and copy?  Feel free to ignore this.

**Review by sstanfield — APPROVED** [review]

**Pull request merged**